dask/rewrite.py

```diff
@@ -0,0 +1,298 @@
+
+
+class RuleSet(object):
+    """A set of rewrite rules.
+
+    Forms a structure for fast rewriting over a set of rewrite rules.
+
+    Parameters:
+    -----------
+    rules: RewriteRule
+        One or instances of RewriteRule
+    """
+
+    def __init__(self, *rules):
+        self._net = Node()
+        self._rules = []
+        for p in rules:
+            self.add(p)
+
+    def add(self, rule):
+        """Add a rule to the RuleSet."""
+
+        if not isinstance(rule, RewriteRule):
+            raise TypeError("rule must be instance of RewriteRule")
+        vars = rule.vars
+        curr_node = self._net
+        ind = len(self._rules)
+        # List of variables, in order they appear in the POT of the term
+        for t in Traverser(rule.lhs):
+            prev_node = curr_node
+            if t in vars:
+                t = VAR
+            if t in curr_node.edges:
+                curr_node = curr_node.edges[t]
+            else:
+                curr_node.edges[t] = Node()
+                curr_node = curr_node.edges[t]
+        # We've reached a leaf node. Add the term index to this leaf.
+        prev_node.edges[t].patterns.append(ind)
+        self._rules.append(rule)
+
+    def iter_matches(self, term):
+        """Lazily find matchings for term from the RuleSet.
+
+        Paramters:
+        ----------
+        term : task
+
+        Returns:
+        --------
+        A generator that yields tuples of ``(rule, subs)``, where ``rule`` is
+        the rewrite rule being matched, and ``subs`` is a dictionary mapping
+        the variables in that lhs of the rule to their matching values in the
+        term."""
+
+        S = Traverser(term)
+        for m, syms in _match(S, self._net):
+            for i in m:
+                rule = self._rules[i]
+                subs = _process_match(rule, syms)
+                if subs is not None:
+                    yield rule, subs
+
+    def _rewrite(self, term):
+        """Apply the rewrite rules in RuleSet to top level of term"""
+
+        for rule, sd in self.iter_matches(term):
+            term = rule.rhs
+            for key, val in sd.items():
+                term = subs(term, key, val)
+            # We use for (...) because it's fast in all cases for getting the
+            # first element from the match iterator. As we only want that
+            # element, we break here
+            break
+        return term
+
+    def rewrite(self, term, strategy="bottom_up"):
+        """Apply the rule set to ``term``.
+
+        Parameters:
+        -----------
+        term: dask, or a task
+            The item to be rewritten
+        strategy: str, optional
+            The rewriting strategy to use. Options are "bottom_up" (default),
+            or "top_level".
+        """
+        func = strategies[strategy]
+        if isinstance(term, dict):
+            return dict((k, func(self, v)) for (k, v) in term.items())
+        return func(self, term)
```

Comment:
I suggest that we keep RuleSets specific to just terms and use other mechanisms (like `toolz.valmap`) to rewrite on many terms within a dict.